### PR TITLE
chore(sca): #PLC-764 Implement scan_diff command

### PR DESCRIPTION
**Pull request by tchaumeny**
Like in SCA scan all, the output_handler is not implemented yet.

### CODE REVIEW

File: tests/unit/cassettes/test_sca_scan_diff.yaml [hunk 1]
```diff
@@ -0,0 +1,194 @@
+interactions:
```
Comment (xblanchot-gg) [inline tests/unit/cassettes/test_sca_scan_diff.yaml:1]: This file reaise an error from the linter in the CI. You should run a pre-commit on it.

**Comment by xblanchot-gg** [comment]
I think we lack some testing for now. We should add an "end-to-end" test from the command with a `CliRunner`. I did not manage to make it work from my shell.
Also all of the the different mods (pre-commit, pre-push, ...) should all be tested once, at least to verify that we get the right content files in each cases in the tar. We can patch the scan calls to avoid them and catch the tar to verify that the expected information is sent to the backend for example.

File: ggshield/cmd/sca/scan.py [hunk 2]
```diff
@@ -183,3 +214,25 @@ def get_sca_scan_all_filepaths(
...
+
+
+def sca_scan_diff(
+    ctx: click.Context,
+    directory: Path,
+    ref: str,
+    include_staged: bool,
+) -> SCAScanDiffOutput:
+    client = SCAClient(ctx.obj["client"])
+    current_ref = INDEX_REF if include_staged else "HEAD"
```
Comment (xblanchot-gg) [inline ggshield/cmd/sca/scan.py:226]: If `current_ref==ref`, we could avoid the scan here I think.

**Review by xblanchot-gg — CHANGES_REQUESTED** [review]

File: ggshield/cmd/sca/scan.py [hunk 3]
```diff
@@ -57,12 +65,35 @@ def scan_diff_cmd(
     ctx: click.Context,
     directory: Optional[Path],
     ref: str,
+    pre_commit: bool,
+    pre_push: bool,
+    pre_receive: bool,
```
[the remark this answers is not in the document]
Comment (tchaumeny) [reply ggshield/cmd/sca/scan.py:70]: I followed what was done on IaC, do we really want to diverge from them on how we handle those flags?

[the remark this answers is not in the document]
Comment (xblanchot-gg) [reply ggshield/cmd/sca/scan.py:70]: Indeed my question was more for Henri as I know it had been discussed after the iac implementation. We need at least these click decorators in the current implem:
```
@click.option(
    "--ref",
    type=click.STRING,
    cls=OptionGroup,
    not_required_if=["pre-commit", "pre-push", "pre-receive"],
    help="A git reference.",
)
@click.option(
    "--pre-commit",
    is_flag=True,
    type=click.STRING,
    cls=OptionGroup,
    not_required_if=["ref", "pre-push", "pre-receive"],
    help="This is an alias for `ggshield iac scan --ref=HEAD --staged`, intended to be used in a pre-commit hook.",
)
@click.option(
    "--pre-push",
    is_flag=True,
    type=click.STRING,
    cls=OptionGroup,
    not_required_if=["ref", "pre-commit", "pre-receive"],
    help="This is an alias for `ggshield iac scan --ref=@{upstream}`, intended to be used in a pre-push hook.",
)
@click.option(
    "--pre-receive",
    is_flag=True,
    type=click.STRING,
    cls=OptionGroup,
    not_required_if=["ref", "pre-commit", "pre-push"],
    help="This is an alias for `ggshield iac scan --ref=HEAD`, intended to be used in a pre-receive hook.",
)
@click.option(
    "--staged",
    is_flag=True,
    help="Whether staged changes should be included into the scan.",
)
```

[the remark this answers is not in the document]
Comment (GG-HH) [reply ggshield/cmd/sca/scan.py:70]: Nope, the options `pre-receive`, `pre-push` etc... where a mistake. We are currently modifying those for IaC.

What we should target looks like:
```
ggshield sca scan -h
Usage: ggshield sca scan [OPTIONS] COMMAND [ARGS]...

Blabla

Options:
  --all
....

Commands:
  ci
  diff
  pre-commit
  pre-push
  pre-receive
  all
```

File: ggshield/cmd/sca/scan.py (hunk 2 above)
Comment (xblanchot-gg) [reply ggshield/cmd/sca/scan.py:226]: I don't think so, that's not really an error from the code point of view. We may print a warning message like "sca scan diff comparing the current state to itself: nothing to do." (message should be better ! )

File: ggshield/cmd/sca/scan.py [hunk 4]
```diff
@@ -57,12 +65,35 @@ def scan_diff_cmd(
...
+    staged: bool,
     **kwargs: Any,
 ) -> int:
     """
     Find SCA vulnerabilities in a git working directory, compared to git history.
```
[the remark this answers is not in the document]
Comment (tchaumeny) [reply ggshield/cmd/sca/scan.py:75]: As discussed, I've only handled the pre-commit case. Once all is stabilized in IaC we will implement other cases, possibly in the same way as they do, clean and comment this.

[the remark this answers is not in the document]
Comment (xblanchot-gg) [reply ggshield/cmd/sca/scan.py:75]: As Henri said, `pre-commit` should now be a command itself, that we run with `ggshield sca scan pre-commit`. So you need to refactor a little bit the command call here.

[the remark this answers is not in the document]
Comment (tchaumeny) [reply ggshield/cmd/sca/scan.py]: Right I updated it.

File: ggshield/cmd/sca/scan.py [hunk 5]
```diff
@@ -53,16 +61,25 @@ def scan_group(*args, **kwargs: Any) -> None:
...
+    Find SCA vulnerabilities in a git working directory, compared to HEAD.
     """
-    return 0
+    if directory is None:
+        directory = Path().resolve()
+
+    # Adds client and required parameters to the context
+    update_context(ctx)
+
+    result = sca_scan_diff(ctx, directory, "HEAD", True)
```
Comment (xblanchot-gg) [inline ggshield/cmd/sca/scan.py:79]: nit: can you use named arguments here ? I'm not fan of having a `True` argument without a name, for readability.

File: ggshield/cmd/sca/scan.py [hunk 6]
```diff
@@ -183,3 +200,28 @@ def get_sca_scan_all_filepaths(
     # Only sca_files field is useful in the case of a full_scan,
     # all the potential files already exist in `all_filepaths`
     return response.sca_files
+
+
+def sca_scan_diff(
```
Comment (xblanchot-gg) [inline ggshield/cmd/sca/scan.py:205]: I think this function can be tested more:
- case of same refs
- empty repository
- repository without sca files (or maybe all SCA files were removed in ref)

Comment (tchaumeny) [reply ggshield/cmd/sca/scan.py:202]: I added a test for the case of same refs. As discussed, I don't see any particular case to be tested for the other cases (that would be testing the backend).
Maybe these should be tested in integration tests though.

**Review by xblanchot-gg — APPROVED** [review]

**Pull request merged**